Merging file outside source directory leads to panic message and stacktrace dump

## Describe the bug

Tried to run merge on file that is not in source directory. Got panic message and long stacktrace dump instead of simple error message

## To reproduce

Run `chezmoi merge *file not in source directory*`

## Expected behavior

Simple error message like `chezmoi: {FILE}: not in source state`

## Output of command with the `--verbose` flag

```console
$ chezmoi merge .zshrc --verbose
panic: .zshrc: not in source state

goroutine 1 [running]:
github.com/twpayne/chezmoi/v2/internal/chezmoi.(*SourceState).MustEntry(0x400027f000?, {{0x40007498b1?, 0x40002ebf00?}})
        /home/builder/.termux-build/chezmoi/build/src/github.com/twpayne/chezmoi/internal/chezmoi/sourcestate.go:800 +0xb0
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).runMergeCmd(0x64ef6f3540?, 0x40000f6900?, {0x40002ebf00?, 0x0?, 0x0?}, 0x0?)
        /home/builder/.termux-build/chezmoi/build/src/github.com/twpayne/chezmoi/internal/cmd/mergecmd.go:50 +0x98
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).newMergeCmd.(*Config).makeRunEWithSourceState.func2(0x40000fb800?, {0x40002ebf00, 0x1, 0x2})
        /home/builder/.termux-build/chezmoi/build/src/github.com/twpayne/chezmoi/internal/cmd/config.go:1528 +0x88
github.com/spf13/cobra.(*Command).execute(0x40000f6900, {0x40002ebee0, 0x2, 0x2})
        /home/builder/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:983 +0x840
github.com/spf13/cobra.(*Command).ExecuteC(0x400056b200)
        /home/builder/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x344
github.com/spf13/cobra.(*Command).Execute(...)
        /home/builder/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).execute(0x400071fd08?, {0x400003c050, 0x3, 0x3})
        /home/builder/.termux-build/chezmoi/build/src/github.com/twpayne/chezmoi/internal/cmd/config.go:1197 +0x70
github.com/twpayne/chezmoi/v2/internal/cmd.runMain({{0x64ed98f120, 0x6}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}}, {0x400003c050, 0x3, ...})
        /home/builder/.termux-build/chezmoi/build/src/github.com/twpayne/chezmoi/internal/cmd/cmd.go:296 +0x180
github.com/twpayne/chezmoi/v2/internal/cmd.Main({{0x64ed98f120, 0x6}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}}, {0x400003c050?, 0x3, ...})
        /home/builder/.termux-build/chezmoi/build/src/github.com/twpayne/chezmoi/internal/cmd/cmd.go:111 +0x60
main.main()
        /home/builder/.termux-build/chezmoi/build/src/github.com/twpayne/chezmoi/main.go:24 +0xd8
```

## Output of `chezmoi doctor`

<details>

```console
$ chezmoi doctor
RESULT    CHECK                       MESSAGE
warning   version                     v2.46.1
ok        latest-version              v2.46.1
ok        os-arch                     android/arm64
ok        uname                       Linux localhost 4.19.191+ #1 SMP PREEMPT Wed Nov 29 21:44:22 CST 2023 aarch64 Android
ok        go-version                  go1.21.6 (gc)
ok        executable                  /data/data/com.termux/files/usr/bin/chezmoi
ok        config-file                 ~/.config/chezmoi/chezmoi.toml, last modified 2024-02-21T09:15:20Z
warning   source-dir                  /sdcard/.dots is a git working tree (dirty)
ok        suspicious-entries          no suspicious entries
warning   working-tree                /sdcard/.dots is a git working tree (dirty)
ok        dest-dir                    ~ is a directory
warning   umask                       077
ok        cd-command                  found /data/data/com.termux/files/usr/bin/zsh
ok        cd-args                     /data/data/com.termux/files/usr/bin/zsh
info      diff-command                not set
ok        edit-command                found /data/data/com.termux/files/usr/bin/micro
ok        edit-args                   micro
ok        git-command                 found /data/data/com.termux/files/usr/bin/git, version 2.43.2
warning   merge-command               vimdiff not found in $PATH
ok        shell-command               found /data/data/com.termux/files/usr/bin/zsh
ok        shell-args                  /data/data/com.termux/files/usr/bin/zsh
info      age-command                 age not found in $PATH
ok        gpg-command                 found /data/data/com.termux/files/usr/bin/gpg2, version 2.4.4
info      pinentry-command            not set
info      1password-command           op not found in $PATH
info      bitwarden-command           bw not found in $PATH
info      bitwarden-secrets-command   bws not found in $PATH
info      dashlane-command            dcli not found in $PATH
info      doppler-command             doppler not found in $PATH
info      gopass-command              gopass not found in $PATH
info      keepassxc-command           keepassxc-cli not found in $PATH
info      keepassxc-db                not set
info      keeper-command              keeper not found in $PATH
info      lastpass-command            lpass not found in $PATH
info      pass-command                pass not found in $PATH
info      passhole-command            ph not found in $PATH
info      rbw-command                 rbw not found in $PATH
info      vault-command               vault not found in $PATH
info      vlt-command                 vlt not found in $PATH
info      secret-command              not set
```

</details>

--- a/internal/cmd/catcmd.go
+++ b/internal/cmd/catcmd.go
@@ -27,9 +27,7 @@ func (c *Config) newCatCmd() *cobra.Command {
 }
 
 func (c *Config) runCatCmd(cmd *cobra.Command, args []string, sourceState *chezmoi.SourceState) error {
-	targetRelPaths, err := c.targetRelPaths(sourceState, args, targetRelPathsOptions{
-		mustBeManaged: true,
-	})
+	targetRelPaths, err := c.targetRelPaths(sourceState, args, nil)
 	if err != nil {
 		return err
 	}

--- a/internal/cmd/chattrcmd.go
+++ b/internal/cmd/chattrcmd.go
@@ -150,9 +150,8 @@ func (c *Config) runChattrCmd(cmd *cobra.Command, args []string, sourceState *ch
 		return err
 	}
 
-	targetRelPaths, err := c.targetRelPaths(sourceState, args[1:], targetRelPathsOptions{
-		mustBeManaged: true,
-		recursive:     c.chattr.recursive,
+	targetRelPaths, err := c.targetRelPaths(sourceState, args[1:], &targetRelPathsOptions{
+		recursive: c.chattr.recursive,
 	})
 	if err != nil {
 		return err

--- a/internal/cmd/config.go
+++ b/internal/cmd/config.go
@@ -631,9 +631,8 @@ func (c *Config) applyArgs(
 			return err
 		}
 	default:
-		targetRelPaths, err = c.targetRelPaths(sourceState, args, targetRelPathsOptions{
-			mustBeManaged: true,
-			recursive:     options.recursive,
+		targetRelPaths, err = c.targetRelPaths(sourceState, args, &targetRelPathsOptions{
+			recursive: options.recursive,
 		})
 		if err != nil {
 			return err
@@ -2437,9 +2436,8 @@ func (c *Config) setEnvironmentVariables() error {
 // sourceAbsPaths returns the source absolute paths for each target path in
 // args.
 func (c *Config) sourceAbsPaths(sourceState *chezmoi.SourceState, args []string) ([]chezmoi.AbsPath, error) {
-	targetRelPaths, err := c.targetRelPaths(sourceState, args, targetRelPathsOptions{
+	targetRelPaths, err := c.targetRelPaths(sourceState, args, &targetRelPathsOptions{
 		mustBeInSourceState: true,
-		mustBeManaged:       true,
 	})
 	if err != nil {
 		return nil, err
@@ -2462,7 +2460,6 @@ func (c *Config) targetRelPath(absPath chezmoi.AbsPath) (chezmoi.RelPath, error)
 
 type targetRelPathsOptions struct {
 	mustBeInSourceState bool
-	mustBeManaged       bool
 	recursive           bool
 }
 
@@ -2471,7 +2468,7 @@ type targetRelPathsOptions struct {
 func (c *Config) targetRelPaths(
 	sourceState *chezmoi.SourceState,
 	args []string,
-	options targetRelPathsOptions,
+	options *targetRelPathsOptions,
 ) (chezmoi.RelPaths, error) {
 	targetRelPaths := make(chezmoi.RelPaths, 0, len(args))
 	for _, arg := range args {
@@ -2484,16 +2481,16 @@ func (c *Config) targetRelPaths(
 			return nil, err
 		}
 		sourceStateEntry := sourceState.Get(targetRelPath)
-		if options.mustBeManaged && sourceStateEntry == nil {
+		if sourceStateEntry == nil {
 			return nil, fmt.Errorf("%s: not managed", arg)
 		}
-		if options.mustBeInSourceState {
+		if options != nil && options.mustBeInSourceState {
 			if _, ok := sourceStateEntry.(*chezmoi.SourceStateRemove); ok {
 				return nil, fmt.Errorf("%s: not in source state", arg)
 			}
 		}
 		targetRelPaths = append(targetRelPaths, targetRelPath)
-		if options.recursive {
+		if options != nil && options.recursive {
 			parentRelPath := targetRelPath
 			// FIXME we should not call s.TargetRelPaths() here - risk of
 			// accidentally quadratic

--- a/internal/cmd/editcmd.go
+++ b/internal/cmd/editcmd.go
@@ -76,9 +76,8 @@ func (c *Config) runEditCmd(cmd *cobra.Command, args []string) error {
 		return err
 	}
 
-	targetRelPaths, err := c.targetRelPaths(sourceState, args, targetRelPathsOptions{
+	targetRelPaths, err := c.targetRelPaths(sourceState, args, &targetRelPathsOptions{
 		mustBeInSourceState: true,
-		mustBeManaged:       true,
 	})
 	if err != nil {
 		return err

--- a/internal/cmd/forgetcmd.go
+++ b/internal/cmd/forgetcmd.go
@@ -28,9 +28,7 @@ func (c *Config) newForgetCmd() *cobra.Command {
 }
 
 func (c *Config) runForgetCmd(cmd *cobra.Command, args []string, sourceState *chezmoi.SourceState) error {
-	targetRelPaths, err := c.targetRelPaths(sourceState, args, targetRelPathsOptions{
-		mustBeManaged: true,
-	})
+	targetRelPaths, err := c.targetRelPaths(sourceState, args, nil)
 	if err != nil {
 		return err
 	}

--- a/internal/cmd/mergecmd.go
+++ b/internal/cmd/mergecmd.go
@@ -37,9 +37,8 @@ func (c *Config) newMergeCmd() *cobra.Command {
 }
 
 func (c *Config) runMergeCmd(cmd *cobra.Command, args []string, sourceState *chezmoi.SourceState) error {
-	targetRelPaths, err := c.targetRelPaths(sourceState, args, targetRelPathsOptions{
+	targetRelPaths, err := c.targetRelPaths(sourceState, args, &targetRelPathsOptions{
 		mustBeInSourceState: true,
-		mustBeManaged:       false,
 		recursive:           true,
 	})
 	if err != nil {

--- a/internal/cmd/readdcmd.go
+++ b/internal/cmd/readdcmd.go
@@ -52,9 +52,8 @@ func (c *Config) runReAddCmd(cmd *cobra.Command, args []string, sourceState *che
 		)
 	} else {
 		var err error
-		targetRelPaths, err = c.targetRelPaths(sourceState, args, targetRelPathsOptions{
-			mustBeManaged: true,
-			recursive:     c.reAdd.recursive,
+		targetRelPaths, err = c.targetRelPaths(sourceState, args, &targetRelPathsOptions{
+			recursive: c.reAdd.recursive,
 		})
 		if err != nil {
 			return err

--- a/internal/cmd/removecmd.go
+++ b/internal/cmd/removecmd.go
@@ -37,9 +37,8 @@ func (c *Config) newRemoveCmd() *cobra.Command {
 }
 
 func (c *Config) runRemoveCmd(cmd *cobra.Command, args []string, sourceState *chezmoi.SourceState) error {
-	targetRelPaths, err := c.targetRelPaths(sourceState, args, targetRelPathsOptions{
-		mustBeManaged: true,
-		recursive:     c.remove.recursive,
+	targetRelPaths, err := c.targetRelPaths(sourceState, args, &targetRelPathsOptions{
+		recursive: c.remove.recursive,
 	})
 	if err != nil {
 		return err

